lib/src/shared/main/java/com/couchbase/lite/internal/core/C4Document.java

```diff
@@ -29,79 +36,195 @@ public static boolean dictContainsBlobs(FLSliceResult dict, FLSharedKeys sk) {
-        byte[] mergeBody, int mergedFlags)
-        throws LiteCoreException;
+    public boolean isSelectedRevFlags(int flag) { return (getSelectedFlags() & flag) == flag; }
 
-    static native void setExpiration(long db, String docID, long timestamp)
-        throws LiteCoreException;
+    //-------------------------------------------------------------------------
+    // protected methods
+    //-------------------------------------------------------------------------
 
-    // - Creating and Updating Documents
+    @SuppressWarnings("NoFinalizer")
+    @Override
+    protected void finalize() throws Throwable {
+        free();
+        super.finalize();
+    }
+
+    //-------------------------------------------------------------------------
+    // package protected methods
+    //-------------------------------------------------------------------------
+
+    int getSelectedFlags() { return withHandle(C4Document::getSelectedFlags, 0); }
+
+    byte[] getSelectedBody() { return withHandle(C4Document::getSelectedBody, null); }
+
+    boolean selectCurrentRevision() { return withHandle(C4Document::selectCurrentRevision, false); }
+
+    void loadRevisionBody() throws LiteCoreException { withHandleVoid(C4Document::loadRevisionBody); }
+
+    boolean hasRevisionBody() { return withHandle(C4Document::hasRevisionBody, false); }
+
+    boolean selectParentRevision() { return withHandle(C4Document::selectParentRevision, false); }
+
+    boolean selectFirstPossibleAncestorOf(String revID) {
+        return withHandle(h -> selectFirstPossibleAncestorOf(h, revID), false);
+    }
+
+    boolean selectNextPossibleAncestorOf(String revID) {
+        return withHandle(h -> selectNextPossibleAncestorOf(h, revID), false);
+    }
+
+    boolean selectCommonAncestorRevision(String revID1, String revID2) {
+        return withHandle(h -> selectCommonAncestorRevision(h, revID1, revID2), false);
+    }
+
+    int purgeRevision(String revID) throws LiteCoreException {
+        return withHandleThrows(h -> purgeRevision(h, revID), 0);
+    }
+
+    @Override
+    void free() {
+        final long hdl;
+        synchronized (lock) {
+            hdl = handle;
+            handle = 0;
+        }
+
+        if (hdl != 0L) { free(hdl); }
+    }
+
+    //-------------------------------------------------------------------------
+    // private methods
+    //-------------------------------------------------------------------------
+
+    private boolean isFlags(int flag) { return (getFlags() & flag) == flag; }
+
+    private boolean conflicted() { return isFlags(C4Constants.DocumentFlags.CONFLICTED); }
+
+    private boolean accessRemoved() { return isSelectedRevFlags(C4Constants.RevisionFlags.PURGED); }
+
+    private <T> T withHandle(Fn.Function<Long, T> fn, T def) {
+        synchronized (lock) {
+            if (handle != 0) { return fn.apply(handle); }
+            Log.w(LogDomain.DATABASE, "Function called on freed object", new Exception());
```

Comment:
What's with `new Exception()` here?

Reply:
The idea was to get a stack trace.  It totally doesn't work.  Going to fix that now.

Reply:
Fixed in Log.  See below.